interfaces/input_parameters/ring.py

```diff
@@ -182,182 +130,348 @@ class Ring:
+            program_shape)
+        self.beta = ring_programs._synchronous_data_program.zeros(
+            program_shape)
+        self.gamma = ring_programs._synchronous_data_program.zeros(
+            program_shape)
+
+        # Populating the synchronous data attribute each row corresponding
+        # to a section
+        for index_section in range(self.n_sections):
+            self.momentum[index_section, :] = momentum_processed[2:]
+            self.beta[index_section, :] = rt.mom_to_beta(
+                momentum_processed[2:], self.Particle.mass)
+            self.gamma[index_section, :] = rt.mom_to_gamma(
+                momentum_processed[2:], self.Particle.mass)
+            self.energy[index_section, :] = rt.momentum_to_energy(
+                momentum_processed[2:], self.Particle.mass)
+            self.kin_energy[index_section, :] = rt.momentum_to_kin_energy(
+                momentum_processed[2:], self.Particle.mass)
+
+        # Extracting and combining the orbit length programs
+        self.section_length = ring_programs.orbit_length_program.combine_single_sections(
+            *[section.length for section in self.Section_list])
+
+        # Reshaping to match the dimensions of the synchronous data program
+        self.section_length = self.section_length.reshape(
+            self.n_sections, self.cycle_time, self.use_turns)
+
+        # Computing the revolution period on the design orbit
+        # as well as revolution frequency and angular frequency
+        self.t_rev = np.dot(self.section_length_design, 1 / (self.beta * c))
+        self.f_rev = 1 / self.t_rev
+        self.omega_rev = 2 * np.pi * self.f_rev
+
+        # Computing the time of flight in each section
+        # and the revolution period on the beam orbit including
+        # possible orbit bumps
+        self.tof_section_orbit = np.array(
+            self.section_length / (self.beta * c))
+        self.t_rev_orbit = np.sum(self.tof_section_orbit, axis=0)
+        self.f_rev_orbit = 1 / self.t_rev_orbit
+        self.omega_rev_orbit = 2 * np.pi * self.f_rev_orbit
+
+        # Recalculating the delta_E
         self.delta_E = np.diff(self.energy, axis=1)
         if self.n_turns > len(self.use_turns):
             self.delta_E = np.zeros(self.energy.shape)
             self._recalc_delta_E()
 
-        self.momentum = self.momentum[2:]
-
-        self.f_rev = 1/self.t_rev
-        self.omega_rev = 2*np.pi*self.f_rev
+        # Determining the momentum compaction orders defined in all sections
+        # The orders 1 and 2 are presently set by default to zeros if
+        # not defined in sections for the calculation of all orders of eta
+        self.alpha_orders = [
+            section.alpha_orders
+            for section in self.Section_list] + [[1], [2]]
 
-        # Momentum compaction, checks, and derived slippage factors
+        self.alpha_orders = np.unique(self.alpha_orders)
```

Comment:
I don't think this will behave correctly for multiple sections.

Reply:
It didn't, but the more recent version works, needs thorough unittests